lib/collection/src/shards/replica_set/clock_set.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Default)]
+pub struct ClockSet {
+    clocks: Vec<Arc<Clock>>,
+}
+
+impl ClockSet {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the first available clock from the set, or create a new one.
+    pub fn get_clock(&mut self) -> ClockGuard {
+        for (id, clock) in self.clocks.iter().enumerate() {
+            if clock.lock() {
+                return ClockGuard::new(id, clock.clone());
+            }
+        }
+
+        let id = self.clocks.len();
+        let clock = Arc::new(Clock::new_locked());
+
+        self.clocks.push(clock.clone());
+
+        ClockGuard::new(id, clock)
+    }
+}
+
+#[derive(Debug)]
+pub struct ClockGuard {
+    id: usize,
+    clock: Arc<Clock>,
+}
+
+impl ClockGuard {
+    fn new(id: usize, clock: Arc<Clock>) -> Self {
+        Self { id, clock }
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    /// Advance clock by a single tick and return current tick.
+    #[must_use = "new clock value must be used"]
+    pub fn tick_once(&mut self) -> u64 {
+        self.clock.tick_once()
+    }
+
+    /// Advance clock to `new_tick`, if `new_tick` is newer than current tick.
+    pub fn advance_to(&mut self, new_tick: u64) {
+        self.clock.advance_to(new_tick)
+    }
+}
+
+impl Drop for ClockGuard {
+    fn drop(&mut self) {
+        self.clock.release();
+    }
+}
+
+#[derive(Debug)]
+struct Clock {
+    /// Tracks the *next* clock tick
+    next_tick: AtomicU64,
+    available: AtomicBool,
+}
+
+impl Clock {
+    fn new_locked() -> Self {
+        Self {
+            next_tick: AtomicU64::new(0),
+            available: AtomicBool::new(false),
+        }
+    }
+
+    /// Advance clock by a single tick and return current tick
+    #[must_use = "new clock tick value must be used"]
+    fn tick_once(&self) -> u64 {
+        // `Clock` tracks *next* tick, so we increment `next_tick` by 1 and return *previous* value
+        // of `next_tick` (which is *exactly* what `fetch_add(1)` does)
+        let current_tick = self.next_tick.fetch_add(1, Ordering::Relaxed);
+
+        // If `current_tick` is `0`, then "revert" `next_tick` back to `0`.
+        // We expect that `advance_to` would be used to advance "past" the initial `0`.
+        //
+        // Executing multiple atomic operations sequentially is not strictly thread-safe,
+        // but we expect that `Clock` would be "locked" before calling `tick_once`.
+        if current_tick == 0 {
+            self.next_tick.store(0, Ordering::Relaxed);
+        }
```

Comment:
> I have another question: do we want additional safeguard around tick 0?
> ...
> We could add some special handling around 0, but it will probably be rather ugly. :/

Disregard that. I've figured out a *very elegant* solution to this.

So, the idea is that `tick_once` only increment `next_tick` if it is already more than `0`. If `next_tick` is `0` we keep returning `0` and don't increment `next_tick`. To "advance" past this "initial `0`" you **have to** use `advance_to`. And because `advance_to` does `new_tick + 1`, even if you call `advance_to(0)` it will still increment `next_tick` to `1` or more.

So, after you restart a node, you have an empty `ClockSet` with all zeroes. Then, *until you've made at least one successful request*, `ClockSet` will stay at `0` (so that operations won't be written into WALs on other nodes with wrong `clock_tick`s).

We always call `advance_to` when collecting/processing responses from other replicas (if response has a `clock_tag`, of course), so once you received the first successful response, the `advance_to` call would update the clock past initial `0` and it will start ticking normally.

Reply:
`tick_once` could be implemented in a more thread-safe manner (using `load` and `compare_and_exchange` in a loop), but we explicitly lock the `Clock` anyway, and updating `next_tick` the way I currently implemented it should be more efficient.